src/emit/expression.rs

```diff
@@ -2120,3 +2122,227 @@ fn runtime_cast<'a>(
+            let llvm_ty = bin.llvm_type(from, ns);
+            let src = bin.build_alloca(function, llvm_ty, "src");
+
+            bin.builder.build_store(src, val.into_int_value());
+
+            let bytes_length: u64 = (*bytes_length).into();
+
+            let dest = call!("__malloc", &[i32_const!(bytes_length).into()])
+                .try_as_basic_value()
+                .left()
+                .unwrap()
+                .into_pointer_value();
+
+            bin.builder.build_call(
+                bin.module.get_function("__leNtobeN").unwrap(),
+                &[src.into(), dest.into(), i32_const!(bytes_length).into()],
+                "",
+            );
+
+            let len = i64_const!(bytes_length);
+
+            (dest, len)
+        }
+        Type::Array(_, dims) => {
+            let to_elem = to.array_elem();
+
+            let to = bin.llvm_type(to, ns);
+
+            let length = match dims.last().unwrap() {
+                ArrayLength::Dynamic => bin.vector_len(val),
+                ArrayLength::Fixed(len) => i32_const!(len.to_u64().unwrap()),
+                _ => unreachable!(),
+            };
+
+            // In Program Runtime v1, we can't do dynamic alloca. Remove the malloc once we move to
+            // program runtime v2
+            let size = bin.builder.build_int_mul(
+                bin.builder.build_int_truncate(
+                    bin.llvm_type(&Type::Slice(Type::Bytes(1).into()), ns)
+                        .size_of()
+                        .unwrap(),
+                    bin.context.i32_type(),
+                    "slice_size",
+                ),
+                length,
+                "size",
+            );
+
+            let output = call!("__malloc", &[size.into()])
+                .try_as_basic_value()
+                .left()
+                .unwrap()
+                .into_pointer_value();
+
+            // loop over seeds
+            let mut builder = LoopBuilder::new(bin, function);
+
+            let index = builder.over(bin, i32_zero!(), length);
+
+            // get value from array
+            let input_elem = bin.array_subscript(from, val.into_pointer_value(), index, ns);
+
+            let from_elem = from.array_elem();
+
+            // If the element is a fixed-length array, do not load it as it's stored in place and not
+            // as a pointer.
+            let load = if let Type::Array(_, dims) = &from_elem {
+                matches!(dims.last(), Some(ArrayLength::Dynamic))
+            } else {
+                true
+            };
+
+            let input_elem = if load {
+                bin.builder
+                    .build_load(bin.llvm_field_ty(&from_elem, ns), input_elem, "elem")
+            } else {
+                input_elem.into()
+            };
+
+            let (data, len) =
+                basic_value_to_slice(bin, input_elem, &from_elem, &to_elem, function, ns);
+
+            let output_data = unsafe {
```

Comment:
[Each unsafe block should have a `SAFETY:` comment ](https://std-dev-guide.rust-lang.org/policy/safety-comments.html)

Reply:
In emit we have `unsafe {}` blocks for all calls to `build_gep()`. This is because llvm can crash if the arguments are not correct, and the "safe" inkwell wrappers do not check for this (in fact I don't think it's possible to fix in inkwell, this should be fixed in llvm).

So I think all the `unsafe {}` blocks in emit are for `build_gep()`. Do we really require `SAFETY: ` comment for each and every one with the same comment (83 occurrences)?

I am not against documenting this - it's certainly useful information - I am just not sure where it should go.

Reply:
Yeah I see. What if we just start doing this from now on?

Reply:
Ok I've added comments. I think it clutters the code

Reply:
`build_gep() is unsafe since it can crash deep in llvm`

This is not the point of the comment. The comment should say why we think this is still safe, despite the `unsafe`. E.g. a proof that the arguments are correct in that case.

>  I think it clutters the code

I got used to it. To the extend I already expect it :grin: